src/Examine.Lucene/Search/LuceneSearchExecutor.cs

```diff
@@ -78,47 +80,119 @@ public ISearchResults Execute()
+                    topDocs = searcher.IndexSearcher.SearchAfter(scoreDocAfter, _luceneQuery, filter, _options.Take, sort, trackDocScores, trackMaxScore);
+                }
+                else if (scoreDocAfter != null && sort == null)
+                {
+                    topDocs = searcher.IndexSearcher.SearchAfter(scoreDocAfter, _luceneQuery, _options.Take);
+                }
+                else
+                {
+                    searcher.IndexSearcher.Search(_luceneQuery, topDocsCollector);
+                    if (sortFields.Length > 0)
+                    {
+                        topDocs = ((TopFieldCollector)topDocsCollector).GetTopDocs(_options.Skip, _options.Take);
+                    }
+                    else
+                    {
+                        topDocs = ((TopScoreDocCollector)topDocsCollector).GetTopDocs(_options.Skip, _options.Take);
+                    }
                 }
 
                 var totalItemCount = topDocs.TotalHits;
 
-                var results = new List<ISearchResult>();
+                var results = new List<ISearchResult>(topDocs.ScoreDocs.Length);
                 for (int i = 0; i < topDocs.ScoreDocs.Length; i++)
                 {
                     var result = GetSearchResult(i, topDocs, searcher.IndexSearcher);
                     results.Add(result);
                 }
+                var searchAfterOptions = GetSearchAfterOptions(topDocs);
+                float maxScore = topDocs.MaxScore;
 
-                return new LuceneSearchResults(results, totalItemCount);
+                return new LuceneSearchResults(results, totalItemCount, maxScore, searchAfterOptions);
             }
         }
 
-        private ISearchResult GetSearchResult(int index, TopDocs topDocs, IndexSearcher luceneSearcher)
+        private static FieldDoc GetScoreDocAfter(LuceneQueryOptions luceneQueryOptions)
+        {
+            FieldDoc scoreDocAfter;
+            var searchAfter = luceneQueryOptions.SearchAfter;
+
+            object[] searchAfterSortFields = new object[0];
+            if (luceneQueryOptions.SearchAfter.Fields != null && luceneQueryOptions.SearchAfter.Fields.Length > 0)
+            {
+                searchAfterSortFields = luceneQueryOptions.SearchAfter.Fields;
+            }
+            if (searchAfter.ShardIndex != null)
+            {
+                scoreDocAfter = new FieldDoc(searchAfter.DocumentId, searchAfter.DocumentScore, searchAfterSortFields, searchAfter.ShardIndex.Value);
+            }
+            else
+            {
+                scoreDocAfter = new FieldDoc(searchAfter.DocumentId, searchAfter.DocumentScore, searchAfterSortFields);
+            }
+
+            return scoreDocAfter;
+        }
+
+        private static SearchAfterOptions GetSearchAfterOptions(TopDocs topDocs)
+        {
+            if (topDocs.TotalHits > 0)
+            {
+                if (topDocs.ScoreDocs.LastOrDefault() is FieldDoc lastFieldDoc && lastFieldDoc != null)
```

Comment:
is ScoreDocs indexed or do we have to use Linq here since that will iterate all of them?

Reply:
ScoreDocs is an array which implements IList. In linq, LastOrDefault() short circuits IList implementations and uses count -1 https://github.com/microsoft/referencesource/blob/master/System.Core/System/Linq/Enumerable.cs. So it won't enumerate the list.

Reply:
Nice, just not sure when that type checking was added to .NET Framework since I feel like older versions didn't always have these checks.